File: apps/app/src/lib/overview-performance.ts
import type { LivelinePoint } from "liveline";

function normalizePoints(
  points: ReadonlyArray<LivelinePoint>,
): LivelinePoint[] {
  const byTime = new Map<number, number>();
  for (const point of points) {
    if (!Number.isFinite(point.time)) continue;
    if (!Number.isFinite(point.value)) continue;
    byTime.set(point.time, point.value);
  }

  return Array.from(byTime.entries())
    .map(([time, value]) => ({ time, value }))
    .sort((a, b) => a.time - b.time);
}

export function forwardFillSeriesOntoBuckets(args: {
  sourcePoints: ReadonlyArray<LivelinePoint>;
  bucketTimesSec: ReadonlyArray<number>;
}): LivelinePoint[] {
  const source = normalizePoints(args.sourcePoints);
  if (source.length === 0 || args.bucketTimesSec.length === 0) return [];

  const points: LivelinePoint[] = [];
  let cursor = 0;
  let lastValue: number | null = null;

  for (const bucketTime of args.bucketTimesSec) {
    while (cursor < source.length && source[cursor]!.time <= bucketTime) {
      const nextValue = source[cursor]!.value;
      lastValue =
        Number.isFinite(nextValue) && nextValue > 0 ? nextValue : lastValue;
      cursor++;
    }

    if (lastValue === null) continue;
    points.push({ time: bucketTime, value: lastValue });
  }

  return points;
}

export function rebaseSeriesFromFirstPoint(
  points: ReadonlyArray<LivelinePoint>,
): LivelinePoint[] {
  const normalized = normalizePoints(points).filter((point) => point.value > 0);
  const baseline = normalized[0]?.value;
  if (!baseline || !Number.isFinite(baseline) || baseline <= 0) return [];

  return normalized.map((point) => ({
    time: point.time,
    value: (point.value / baseline) * 100,
  }));
}

export function getPointValueAtTime(
  points: ReadonlyArray<LivelinePoint>,
  targetTime: number,
): number | null {
  const normalized = normalizePoints(points);
  if (normalized.length === 0) return null;

  let low = 0;
  let high = normalized.length - 1;

  while (low <= high) {
    const mid = Math.floor((low + high) / 2);
    const midPoint = normalized[mid];
    if (!midPoint) break;
    if (midPoint.time === targetTime) return midPoint.value;
    if (midPoint.time < targetTime) low = mid + 1;
    else high = mid - 1;
  }

  const right = normalized[low];
  const left = normalized[low - 1];
  if (!left && !right) return null;
  if (!left) return right?.value ?? null;
  if (!right) return left.value;

  return Math.abs(left.time - targetTime) <= Math.abs(right.time - targetTime)
    ? left.value
    : right.value;
}

export interface RebasedComparisonResult {
  baselineTime: number | null;
  portfolioPoints: LivelinePoint[];
  marketPoints: LivelinePoint[];
  portfolioReturnPct: number | null;
  marketReturnPct: number | null;
  outperformancePct: number | null;
}

export function buildRebasedComparison(args: {
  portfolioPoints: ReadonlyArray<LivelinePoint>;
  marketPoints: ReadonlyArray<LivelinePoint>;
}): RebasedComparisonResult {
  const portfolio = normalizePoints(args.portfolioPoints).filter(
    (point) => point.value > 0,
  );
  const market = normalizePoints(args.marketPoints).filter(
    (point) => point.value > 0,
  );
  if (portfolio.length === 0 || market.length === 0) {
    return {
      baselineTime: null,
      portfolioPoints: [],
      marketPoints: [],
      portfolioReturnPct: null,
      marketReturnPct: null,
      outperformancePct: null,
    };
  }

  const marketByTime = new Map(
    market.map((point) => [point.time, point.value] as const),
  );
  const portfolioByTime = new Map(
    portfolio.map((point) => [point.time, point.value] as const),
  );

  const baselinePortfolioPoint = portfolio.find((point) => {
    const marketValue = marketByTime.get(point.time);
    return (
      typeof marketValue === "number" &&
      Number.isFinite(marketValue) &&
      marketValue > 0
    );
  });

  if (!baselinePortfolioPoint) {
    return {
      baselineTime: null,
      portfolioPoints: [],
      marketPoints: [],
      portfolioReturnPct: null,
      marketReturnPct: null,
      outperformancePct: null,
    };
  }

  const baselineTime = baselinePortfolioPoint.time;
  const baselinePortfolioValue = baselinePortfolioPoint.value;
  const baselineMarketValue = marketByTime.get(baselineTime) ?? null;
  if (
    baselineMarketValue === null ||
    !Number.isFinite(baselineMarketValue) ||
    baselinePortfolioValue <= 0 ||
    baselineMarketValue <= 0
  ) {
    return {
      baselineTime: null,
      portfolioPoints: [],
      marketPoints: [],
      portfolioReturnPct: null,
      marketReturnPct: null,
      outperformancePct: null,
    };
  }

  const commonTimes = portfolio
    .map((point) => point.time)
    .filter((time) => time >= baselineTime && marketByTime.has(time));

  const portfolioPoints = commonTimes.map((time) => ({
    time,
    value:
      ((portfolioByTime.get(time) ?? baselinePortfolioValue) /
        baselinePortfolioValue) *
      100,
  }));
  const marketPoints = commonTimes.map((time) => ({
    time,
    value:
      ((marketByTime.get(time) ?? baselineMarketValue) / baselineMarketValue) *
      100,
  }));

  const latestPortfolio =
    portfolioPoints[portfolioPoints.length - 1]?.value ?? null;
  const latestMarket = marketPoints[marketPoints.length - 1]?.value ?? null;

  const portfolioReturnPct =
    latestPortfolio !== null && Number.isFinite(latestPortfolio)
      ? latestPortfolio - 100
      : null;
  const marketReturnPct =
    latestMarket !== null && Number.isFinite(latestMarket)
      ? latestMarket - 100
      : null;
  const outperformancePct =
    portfolioReturnPct !== null && marketReturnPct !== null
      ? portfolioReturnPct - marketReturnPct
      : null;

  return {
    baselineTime,
    portfolioPoints,
    marketPoints,
    portfolioReturnPct,
    marketReturnPct,
    outperformancePct,
  };
}

File: apps/ios/Packages/AggrCore/Sources/AggrCore/Overview/OverviewPerformance.swift
import Foundation

/// Port of `lib/overview-performance.ts`.
public enum OverviewPerformance {
  static func normalize(_ points: [TimePoint]) -> [TimePoint] {
    var byTime: [Int: Double] = [:]
    for p in points where p.value.isFinite { byTime[p.epochSeconds] = p.value }
    return byTime.keys.sorted().map { TimePoint(epochSeconds: $0, value: byTime[$0]!) }
  }

  /// Forward-fill sparse source points onto denser bucket times (positive values only carry).
  public static func forwardFill(source: [TimePoint], bucketTimesSec: [Int]) -> [TimePoint] {
    let src = normalize(source)
    guard !src.isEmpty, !bucketTimesSec.isEmpty else { return [] }
    var out: [TimePoint] = []
    var cursor = 0
    var last: Double? = nil
    for t in bucketTimesSec {
      while cursor < src.count && src[cursor].epochSeconds <= t {
        let v = src[cursor].value
        if v.isFinite && v > 0 { last = v }
        cursor += 1
      }
      if let last { out.append(TimePoint(epochSeconds: t, value: last)) }
    }
    return out
  }

  public static func rebaseFromFirstPoint(_ points: [TimePoint]) -> [TimePoint] {
    let n = normalize(points).filter { $0.value > 0 }
    guard let baseline = n.first?.value, baseline.isFinite, baseline > 0 else { return [] }
    return n.map { TimePoint(epochSeconds: $0.epochSeconds, value: $0.value / baseline * 100) }
  }

  /// Nearest point value (binary search; ties prefer the left neighbour).
  public static func valueAt(_ points: [TimePoint], time target: Int) -> Double? {
    let n = normalize(points)
    guard !n.isEmpty else { return nil }
    var lo = 0, hi = n.count - 1
    while lo <= hi {
      let mid = (lo + hi) / 2
      if n[mid].epochSeconds == target { return n[mid].value }
      if n[mid].epochSeconds < target { lo = mid + 1 } else { hi = mid - 1 }
    }
    let right = lo < n.count ? n[lo] : nil
    let left = lo - 1 >= 0 ? n[lo - 1] : nil
    switch (left, right) {
    case (nil, nil): return nil
    case (nil, let r?): return r.value
    case (let l?, nil): return l.value
    case (let l?, let r?): return abs(l.epochSeconds - target) <= abs(r.epochSeconds - target) ? l.value : r.value
    }
  }

  public struct RebasedComparison: Sendable, Hashable {
    public var baselineTime: Int?
    public var portfolioPoints: [TimePoint]
    public var marketPoints: [TimePoint]
    public var portfolioReturnPct: Double?
    public var marketReturnPct: Double?
    public var outperformancePct: Double?
    public static let empty = RebasedComparison(baselineTime: nil, portfolioPoints: [], marketPoints: [], portfolioReturnPct: nil, marketReturnPct: nil, outperformancePct: nil)
  }

  /// Rebase both series to 100 at the first overlapping valid point.
  public static func buildRebasedComparison(portfolio: [TimePoint], market: [TimePoint]) -> RebasedComparison {
    let p = normalize(portfolio).filter { $0.value > 0 }
    let m = normalize(market).filter { $0.value > 0 }
    guard !p.isEmpty, !m.isEmpty else { return .empty }
    let marketByTime = Dictionary(m.map { ($0.epochSeconds, $0.value) }, uniquingKeysWith: { a, _ in a })
    let portfolioByTime = Dictionary(p.map { ($0.epochSeconds, $0.value) }, uniquingKeysWith: { a, _ in a })
    guard let baseP = p.first(where: { (marketByTime[$0.epochSeconds] ?? 0) > 0 }) else { return .empty }
    let baselineTime = baseP.epochSeconds
    let baseM = marketByTime[baselineTime]!
    guard baseP.value > 0, baseM > 0 else { return .empty }
    let common = p.filter { $0.epochSeconds >= baselineTime && marketByTime[$0.epochSeconds] != nil }.map(\.epochSeconds)
    let pp = common.map { TimePoint(epochSeconds: $0, value: (portfolioByTime[$0] ?? baseP.value) / baseP.value * 100) }
    let mp = common.map { TimePoint(epochSeconds: $0, value: (marketByTime[$0] ?? baseM) / baseM * 100) }
    let pr = pp.last.map { $0.value - 100 }
    let mr = mp.last.map { $0.value - 100 }
    let out: Double? = (pr != nil && mr != nil) ? pr! - mr! : nil
    return RebasedComparison(baselineTime: baselineTime, portfolioPoints: pp, marketPoints: mp, portfolioReturnPct: pr, marketReturnPct: mr, outperformancePct: out)
  }

  /// Nearest series time to a target (for scrub → onHover). Ties prefer the left neighbour.
  public static func closestTime(_ points: [TimePoint], to target: Int) -> Int? {
    guard !points.isEmpty else { return nil }
    let sorted = points.sorted { $0.epochSeconds < $1.epochSeconds }
    var lo = 0, hi = sorted.count - 1
    while lo <= hi {
      let mid = (lo + hi) / 2
      if sorted[mid].epochSeconds == target { return target }
      if sorted[mid].epochSeconds < target { lo = mid + 1 } else { hi = mid - 1 }
    }
    let right = lo < sorted.count ? sorted[lo].epochSeconds : nil
    let left = lo - 1 >= 0 ? sorted[lo - 1].epochSeconds : nil
    switch (left, right) {
    case (nil, nil): return nil
    case (nil, let r?): return r
    case (let l?, nil): return l
    case (let l?, let r?): return abs(l - target) <= abs(r - target) ? l : r
    }
  }
}
